python/lsst/ts/mtmount/telemetry_client.py

```diff
@@ -214,65 +208,46 @@ async def start(self):
         if self.connected:
             raise RuntimeError("Already connected")
         try:
-            connect_coro = asyncio.open_connection(host=self.host, port=self.port)
-            self.reader, self.writer = await asyncio.wait_for(
-                connect_coro, timeout=self.connection_timeout
-            )
+            await asyncio.wait_for(super().start(), timeout=self.connection_timeout)
             await self.controller.evt_telemetryConnected.set_write(
                 connected=self.connected
             )
         except Exception as e:
-            err_msg = f"Could not open connection to host={self.host}, port={self.port}"
-            self.log.exception(err_msg)
-            self.done_task.set_exception(e)
+            self.log.exception(
+                f"Could not open connection to host={self.host}, port={self.port}: {e!r}"
+            )
+            if not self.done_task.done():
+                self.done_task.set_exception(e)
             return
 
-        self.read_task = asyncio.create_task(self.read_loop())
-        self.slowdown_detection_loop_task = asyncio.create_task(
-            self.slowdown_detection_loop()
-        )
-        self.log.info("running")
-
     def signal_handler(self):
         """Handle signals such as SIGTERM."""
         self.log.info("signal_handler")
-        self.start_task.cancel()
+        self.heartbeat_task.cancel()
         self.read_task.cancel()
         self.next_clock_offset_task.cancel()
-        self.slowdown_detection_loop_task.cancel()
         self.controller.salinfo.basic_close()
         self.controller.salinfo.domain.basic_close()
-        writer = self.writer
-        self.reader = None
-        self.writer = None
-        if writer:
-            writer.close()
-        if not self.done_task.done():
-            self.done_task.set_result(None)
-
-    async def close(self):
-        """Disconnect from the TCP/IP controller."""
-        self.log.info("disconnecting")
-        self.start_task.cancel()
-        self.read_task.cancel()
-        self.next_clock_offset_task.cancel()
-        self.slowdown_detection_loop_task.cancel()
-        writer = self.writer
-        self.reader = None
-        self.writer = None
-        if writer:
-            writer.close()
-            try:
-                await asyncio.wait_for(writer.wait_closed(), timeout=1)
-            except asyncio.TimeoutError:
+        asyncio.create_task(self.close())
+
+    async def connect_callback(self, client):
+        await self.controller.evt_telemetryConnected.set_write(connected=self.connected)
+        if self.connected:
+            if hasattr(self.controller, "tel_telemetryClientHeartbeat"):
```

Comment:
Can you use the `self.has_heartbeat_topic` attribute instead?

Reply:
Thanks for catching this. I intended to delete the has_heartbeat_topic in order to the detection next to the code that needs the information. (If the detection needed to be done in a loop then I would prefer an attribute, but in this case the hasattr is only called when a new connection is made).